librz/util/structured_data.c

```diff
@@ -0,0 +1,576 @@
+ *
+ * \param      parent  Where to add the child RzStructuredData
+ *
+ * \return     On success returns a valid pointer to the child RzStructuredData, otherwise NULL
+ */
+RZ_API RZ_BORROW RzStructuredData *rz_structured_data_array_add_map(RZ_NONNULL RzStructuredData *parent) {
+	rz_return_val_if_fail(parent && parent->type == STRUCTURED_DATA_TYPE_ARRAY, false);
+
+	RzStructuredData *child = structured_data_new_map();
+	if (!structured_data_array_add(parent, child)) {
+		return NULL;
+	}
+
+	return child;
+}
+
+/**
+ * \brief      Returns a child, array type, RzStructuredData that has been added to the given parent
+ *
+ * \param      parent  Where to add the child RzStructuredData
+ *
+ * \return     On success returns a valid pointer to the child RzStructuredData, otherwise NULL
+ */
+RZ_API RZ_BORROW RzStructuredData *rz_structured_data_array_add_array(RZ_NONNULL RzStructuredData *parent) {
+	rz_return_val_if_fail(parent && parent->type == STRUCTURED_DATA_TYPE_ARRAY, false);
+
+	RzStructuredData *child = structured_data_new_array();
+	if (!structured_data_array_add(parent, child)) {
+		return NULL;
+	}
+
+	return child;
+}
+
+/**
+ * \brief      Adds an owned RzStructuredData child to an array type parent
+ *
+ * \param      parent  Where to add the child RzStructuredData
+ * \param      child   The child RzStructuredData
+ *
+ * \return     On success returns true, otherwise the child gets deallocated by the function and return false.
+ */
+RZ_API bool rz_structured_data_array_add(RZ_NONNULL RzStructuredData *parent, RZ_NONNULL RZ_OWN RzStructuredData *child) {
+	rz_return_val_if_fail(parent && parent->type == STRUCTURED_DATA_TYPE_ARRAY && child, false);
+
+	return structured_data_array_add(parent, child);
+}
+
+/**
+ * \brief      Adds unsigned type child to an array type parent
+ *
+ * \param      parent  Where to add the child RzStructuredData
+ * \param      n       The value to assign to the child
+ * \param      hex     When true the type is going to be set as base16 number instead of base10 unsigned number.
+ *
+ * \return     On success returns true, otherwise the child gets deallocated by the function and return false.
+ */
+RZ_API bool rz_structured_data_array_add_unsigned(RZ_NONNULL RzStructuredData *parent, ut64 n, bool hex) {
+	rz_return_val_if_fail(parent && parent->type == STRUCTURED_DATA_TYPE_ARRAY, false);
+
+	RzStructuredData *child = structured_data_new_unsigned(n, hex);
+	return structured_data_array_add(parent, child);
+}
+
+/**
+ * \brief      Adds signed type child to an array type parent
+ *
+ * \param      parent  Where to add the child RzStructuredData
+ * \param      n       The value to assign to the child
+ * \param      hex     When true the type is going to be set as base16 number instead of base10 unsigned number.
+ *
+ * \return     On success returns true, otherwise the child gets deallocated by the function and return false.
```

Comment:
```suggestion
 * \return     On success returns true, otherwise false.
```

The deallocation part isn't visible to the user. SO doesn't need to be documented. Also below.

Reply:
i disagree. this is a usage issue where the ownership gets passed to the function, and i think it should be noted in the comments.

Reply:
If the child is given via an `RZ_OWN` pointer. Yes. But for primitive types it doesn't make sense to talk about deallocation, doesn't it?

Reply:
Oh, i just noticed that i copied and pasted this on the other add_* types... sorry my bad. this requested change is valid.